modules/core/src/smithy4s/codecs/FieldSkipCompiler.scala

```diff
@@ -0,0 +1,144 @@
+/*
+ *  Copyright 2021-2025 Disney Streaming
+ *
+ *  Licensed under the Tomorrow Open Source Technology License, Version 1.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *     https://disneystreaming.github.io/TOST-1.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package smithy4s.codecs
+
+import smithy4s.schema.Field
+import smithy4s.schema.Schema
+
+trait FieldSkipCompiler { self =>
+  def compile[A](
+      field: Field[?, A]
+  ): FieldSkipCompiler.ShouldSkip[A]
+
+  def combine(
+      other: FieldSkipCompiler
+  ): FieldSkipCompiler =
+    new FieldSkipCompiler {
+
+      def compile[A](
+          field: Field[?, A]
+      ): FieldSkipCompiler.ShouldSkip[A] = {
+        val r1 = self.compile(field)
+        val r2 = other.compile(field)
+        a => r1(a) || r2(a)
+      }
+    }
+}
+
+object FieldSkipCompiler {
+
+  type ShouldSkip[A] = A => Boolean
+
+  private trait SkipNonRequired extends FieldSkipCompiler {
+
+    final def compile[A](
+        field: Field[?, A]
+    ): FieldSkipCompiler.ShouldSkip[A] = {
+      if (field.isRequired) Function.const(false)
+      else compileOptional(field)
+
+    }
+
+    def compileOptional[A](
```

Comment:
nitpick: I'm not too happy about this having `Optional` but not having an Option in the signature, but since it's private I guess it's fine

Reply:
me neither